### ISSUE TITLE
Proposal: Permit empty arrays

### ISSUE THREAD

**Issue by tylerchr** [tylerchr/jstn#5]
This is a proposal to permit the representation of arrays with no child element type in the JSTN spec. That is, this spec makes `[]` a valid JSTN text. Currently the spec indicates that

> An array structure is represented as a pair of square brackets surrounding a single type declaration.

This implies that it is not permitted to have an array with _no_ type declaration, such as the example above. I propose that we revise this to permit the omission of a type declaration inside the array brackets, with the interpretation that the array MUST contain exactly zero elements.

The rationale for this change is as follows:

1. It expands the expressiveness of JSTN at very little cost to JSTN users or spec implementors.
2. It achieves a certain symmetry with the behavior of object types, which permit the object to contain zero properties.
3. In Go, it makes meaningful all possible configurations of a [Type](https://godoc.org/github.com/tylerchr/jstn#Type) struct with `Kind: Array` (currently, such a Type with `Items: nil` is of undefined validity).

**Comment by davejhilton** [comment]
In my mind, this falls under the special category of things that both:

1. make perfect, logical sense in theory... and yet...
2. are almost certainly never that useful in real life (what are some conceivable cases for wanting to **_require_** a property to be an **_empty_** array?)

Generally, when things fall into that special category... (e.g., "Just because you _can_ doesn't mean you _should_) I question the value of spending time to implement them.

UNLESS—as is the case here—there are meaningful benefits to doing so, and it would end up simplifying the spec, the grammar, and the implementation(s).

So I have no qualms with accepting this proposal.

👍

**Comment by tylerchr** [comment]
You make a fair point. I guess I didn't make this clear in the original comment, but the root reason that led me down this road was simply that it was surprising that it doesn't work. I encountered it when coding up the generator: I intuitively assumed the child type as optional ("If `{}` is valid, then `[]` probably is too"), only to realize that wasn't defined behavior when I went to write a test and realized I couldn't parse such a value.

From that perspective, this proposal is to support some syntax that, while perhaps useless, intuitively seems like it ought to at least parse. I think your comment explains concisely why that might not always be good justification, but that this is a reasonable exception considering the low cost.

In any case, I'm going to accept this and eventually submit a PR.
